### PR TITLE
mediainfo and  ffmpeg tests

### PR DESCRIPTION
**Pull request by stasinos**
Towards issue #1651

### CODE REVIEW

**Comment by SubJunk** [comment]
That seems similar to what I tried when I tried this. If I remember right, I was getting a NPE on getFilename. I would push what I had but I'm on a different computer right now. Anyway this seems like a good idea.
One change I would possible make would be to call the file FFmpegVideo since there is a lot of setup involved but the objective is to test FFmpegVideo.launchTranscode

**Comment by stasinos** [comment]
@SubJunk I see PMS.killOld() called in the main() function when UMS starts up, but I cannot use PMS.killOld() when setting a testcase because it is private.
Things seem to work anyway, but would it be useful to make it public so it can be called by tests?

**Comment by stasinos** [comment]
There is a weird difference in how mediainfo get populated on Travis and on my PC. Here I get:

mediainfo: Container: MP4, Size: 9441436, Video Bitrate: 5016576, Video Tracks: 1, Video Codec: h264, Duration: 0:00:15.420, Video Resolution: 1920 x 1080, Frame Rate: 29.97, Mime Type: video/mp4

But on Travis only:

mediainfo: Size: 9441436, Mime Type: video/mpeg

which causes pretty much all tests to fail.

The only visible difference is that on Travis I get:
java.io.FileNotFoundException: /home/travis/.config/UMS-tests/UMS.cred (No such file or directory)
when calling PMS.getConfiguration().initCred()

Could these be realted?

**Comment by Nadahar** [comment]
@stasinos Yes, I think they are. A lot of things in the code relies on `PMS.getConfiguration()` being initialized. You can initialize a default instance by passing `false` for "load file". If you look in some of the other test classes you will see how it is done.

**Comment by Nadahar** [comment]
@stasinos Sorry, the above was incorrect. This class initializes the configuration, but also does some other strange things. I should check the code properly before answering :wink:

**Comment by Nadahar** [comment]
The reason for the stack track shown in the log is the change you made in eee311680e584e81d5420d8843ec35d5bcf7baeb.

There won't be any credential file on Travis, so you probably shouldn't log the stack trace for that. I don't know why you want to initialize the credentials in the first place though, but this probably isn't related then.

**Comment by Nadahar** [comment]
I suspect that the reason it fails on Travis is that `pexels-video-4809.mp4` isn't found/loaded, and thus no container is parsed.

**Comment by stasinos** [comment]
No, the file is there and found, I checked already. See commit 69026c1 .

**Comment by SubJunk** [comment]
@stasinos I don't have a problem with making `killOld` public :) Nice work in here, this is really coming along. By the way we have a video testing suite that I put together a while ago that might be useful https://sourceforge.net/projects/unimediaserver/files/Official%20Releases/Windows/VideoTestingSuite-1.0.7z/download

We can also add subtitles to those files when we want to create more test cases

**Comment by SubJunk** [comment]
@stasinos I get the same error as Travis when I run it locally on macOS

**Comment by SubJunk** [comment]
@stasinos I have fixed it by replacing lines 73-74 with:

```
		dlna.getParent().setDefaultRenderer(RendererConfiguration.getDefaultConf());
		dlna.resolveFormat();
		dlna.syncResolve();
```

**Comment by stasinos** [comment]
Yay! thx @SubJunk 
Now only MacOS is failing

**Comment by stasinos** [comment]
> By the way we have a video testing suite that I put together a while ago that might be useful ..

@SubJunk Shouldn't this be committed? Otherwise it cannot be used for automatic testing. Is this clip clear for re-distributed? The clip I committed is from pexels, free to modify and re-use, cf. https://videos.pexels.com/video-license

File: src/test/java/net/pms/dlna/DLNAMediaInfoTest.java [hunk 1]
```diff
@@ -0,0 +1,152 @@
...
+
+		// Create the PMS instance returned by get()
+		PMS.get();
+		
+		// Create handles to the test content
+		// This comes from RequestV2::answer()
+
+		DLNAResource parent = new VirtualFolder("test","test");
+		parent.setDefaultRenderer(RendererConfiguration.getDefaultConf());
+		PMS.getGlobalRepo().add(parent);
```
Comment (valib) [inline src/test/java/net/pms/dlna/DLNAMediaInfoTest.java]: @stasinos I have the same problem like Travis. The GlobalRepo is not initialized and is `null`. Tested with Eclipse and WIN10

Comment (stasinos) [reply src/test/java/net/pms/dlna/DLNAMediaInfoTest.java]: Are you running the complete UMS test suite or this particular test only?
My current understanding of the issue is that if I run this test only, then all is well. But I do a complete mvn validate, then some other test has initialized PMS and my invokation of PMS.get() has no effect; but for some reason the global repo has disappeared in the meantime.

Comment (valib) [reply src/test/java/net/pms/dlna/DLNAMediaInfoTest.java]: @stasinos I am running all tests.

```
[INFO] Scanning for projects...
[INFO] 
[INFO] ----------------------------< net.pms:ums >-----------------------------
[INFO] Building Universal Media Server 7.6.1-SNAPSHOT
[INFO] --------------------------------[ jar ]---------------------------------
[INFO] 
[INFO] --- maven-enforcer-plugin:1.4.1:enforce (enforce) @ ums ---
[INFO] 
[INFO] --- git-commit-id-plugin:2.2.3:revision (default) @ ums ---
[INFO] 
[INFO] --- maven-resources-plugin:2.6:resources (default-resources) @ ums ---
[INFO] Using 'UTF-8' encoding to copy filtered resources.
[INFO] Copying 429 resources to resources
[INFO] Copying 1 resource to resources
[INFO] Copying 1 resource to ..
[INFO] 
[INFO] --- maven-antrun-plugin:1.8:run (process-resources-windows) @ ums ---
[INFO] Executing tasks

main:
[INFO] Executed tasks
[INFO] 
[INFO] --- maven-compiler-plugin:3.7.0:compile (default-compile) @ ums ---
[INFO] Nothing to compile - all classes are up to date
[INFO] 
[INFO] --- maven-resources-plugin:2.6:testResources (default-testResources) @ ums ---
[INFO] Using 'UTF-8' encoding to copy filtered resources.
[INFO] Copying 1 resource to resources
[INFO] Copying 15 resources to net/pms/util
[INFO] Copying 2 resources to net/pms/dlna
[INFO] 
[INFO] --- maven-compiler-plugin:3.7.0:testCompile (default-testCompile) @ ums ---
[INFO] Changes detected - recompiling the module!
[INFO] Compiling 40 source files to E:\Workspace\UniversalMediaServer\target\test-classes
[INFO] 
[INFO] --- maven-surefire-plugin:2.20:test (default-test) @ ums ---
[INFO] 
[INFO] -------------------------------------------------------
[INFO]  T E S T S
[INFO] -------------------------------------------------------
[INFO] Running net.pms.configuration.FormatConfigurationRegressionTest
[INFO] Tests run: 107, Failures: 0, Errors: 0, Skipped: 0, Time elapsed: 0.143 s - in net.pms.configuration.FormatConfigurationRegressionTest
[INFO] Running net.pms.configuration.PmsConfigurationTest
[INFO] Tests run: 2, Failures: 0, Errors: 0, Skipped: 0, Time elapsed: 2.65 s - in net.pms.configuration.PmsConfigurationTest
[INFO] Running net.pms.configuration.RendererConfigurationTest
[INFO] Tests run: 3, Failures: 0, Errors: 0, Skipped: 0, Time elapsed: 0.872 s - in net.pms.configuration.RendererConfigurationTest
[INFO] Running net.pms.database.TableFilesStatusTest
[INFO] Tests run: 2, Failures: 0, Errors: 0, Skipped: 0, Time elapsed: 0.41 s - in net.pms.database.TableFilesStatusTest
[INFO] Running net.pms.dlna.DLNAMediaAudioTest
[INFO] Tests run: 2, Failures: 0, Errors: 0, Skipped: 0, Time elapsed: 0 s - in net.pms.dlna.DLNAMediaAudioTest
[INFO] Running net.pms.dlna.DLNAMediaInfoTest
[ERROR] Tests run: 1, Failures: 0, Errors: 1, Skipped: 0, Time elapsed: 0.035 s <<< FAILURE! - in net.pms.dlna.DLNAMediaInfoTest
[ERROR] net.pms.dlna.DLNAMediaInfoTest  Time elapsed: 0.035 s  <<< ERROR!
java.lang.NullPointerException
	at net.pms.dlna.DLNAMediaInfoTest.SetUPClass(DLNAMediaInfoTest.java:76)

[INFO] Running net.pms.dlna.DLNAMediaSubtitleTest
[INFO] Tests run: 8, Failures: 0, Errors: 0, Skipped: 0, Time elapsed: 0.166 s - in net.pms.dlna.DLNAMediaSubtitleTest
[INFO] Running net.pms.dlna.LibMediaInfoParserTest
[INFO] Tests run: 11, Failures: 0, Errors: 0, Skipped: 0, Time elapsed: 0.016 s - in net.pms.dlna.LibMediaInfoParserTest
[INFO] Running net.pms.formats.FormatsExtensionTest
[INFO] Tests run: 1, Failures: 0, Errors: 0, Skipped: 0, Time elapsed: 0.031 s - in net.pms.formats.FormatsExtensionTest
[INFO] Running net.pms.formats.FormatsRegressionTest
[INFO] Tests run: 62, Failures: 0, Errors: 0, Skipped: 0, Time elapsed: 0.027 s - in net.pms.formats.FormatsRegressionTest
[INFO] Running net.pms.formats.v2.AudioAttributeTest
[INFO] Tests run: 5, Failures: 0, Errors: 0, Skipped: 0, Time elapsed: 0 s - in net.pms.formats.v2.AudioAttributeTest
[INFO] Running net.pms.formats.v2.AudioPropertiesTest
[INFO] Tests run: 13, Failures: 0, Errors: 0, Skipped: 0, Time elapsed: 0.007 s - in net.pms.formats.v2.AudioPropertiesTest
[INFO] Running net.pms.formats.v2.SubtitleTypeTest
[INFO] Tests run: 17, Failures: 0, Errors: 0, Skipped: 0, Time elapsed: 0.027 s - in net.pms.formats.v2.SubtitleTypeTest
[INFO] Running net.pms.formats.v2.SubtitleUtilsTest
[INFO] Tests run: 5, Failures: 0, Errors: 0, Skipped: 0, Time elapsed: 0.032 s - in net.pms.formats.v2.SubtitleUtilsTest
[INFO] Running net.pms.logging.LoggingTest
DEBUG 18:22:32.438 [main] LogBack started with configuration file: E:\Workspace\UniversalMediaServer\src\main\external-resources\logback.headless.xml
[INFO] Tests run: 3, Failures: 0, Errors: 0, Skipped: 0, Time elapsed: 0.134 s - in net.pms.logging.LoggingTest
[INFO] Running net.pms.newgui.components.SpinnerIntModelTest
[INFO] Tests run: 1, Failures: 0, Errors: 0, Skipped: 0, Time elapsed: 0.001 s - in net.pms.newgui.components.SpinnerIntModelTest
[INFO] Running net.pms.newgui.components.TextAreaFIFOTest
[INFO] Tests run: 1, Failures: 0, Errors: 0, Skipped: 0, Time elapsed: 0.219 s - in net.pms.newgui.components.TextAreaFIFOTest
[INFO] Running net.pms.newgui.LanguageSelectionTest
[INFO] Tests run: 2, Failures: 0, Errors: 0, Skipped: 0, Time elapsed: 0.02 s - in net.pms.newgui.LanguageSelectionTest
[INFO] Running net.pms.newgui.ViewLevelTest
[INFO] Tests run: 1, Failures: 0, Errors: 0, Skipped: 0, Time elapsed: 0 s - in net.pms.newgui.ViewLevelTest
[INFO] Running net.pms.test.formats.FormatFactoryTest
[INFO] Tests run: 2, Failures: 0, Errors: 0, Skipped: 0, Time elapsed: 0.036 s - in net.pms.test.formats.FormatFactoryTest
[INFO] Running net.pms.test.formats.FormatRecognitionTest
[INFO] Tests run: 6, Failures: 0, Errors: 0, Skipped: 0, Time elapsed: 0.222 s - in net.pms.test.formats.FormatRecognitionTest
[INFO] Running net.pms.test.formats.FormatTest
[INFO] Tests run: 2, Failures: 0, Errors: 0, Skipped: 0, Time elapsed: 0 s - in net.pms.test.formats.FormatTest
[INFO] Running net.pms.test.Iso639Test
[INFO] Tests run: 1, Failures: 0, Errors: 0, Skipped: 0, Time elapsed: 0.003 s - in net.pms.test.Iso639Test
[INFO] Running net.pms.test.TaskRunnerTest
[INFO] Tests run: 2, Failures: 0, Errors: 0, Skipped: 0, Time elapsed: 1.005 s - in net.pms.test.TaskRunnerTest
[INFO] Running net.pms.util.FilePermissionsTest
[INFO] Tests run: 4, Failures: 0, Errors: 0, Skipped: 0, Time elapsed: 0.023 s - in net.pms.util.FilePermissionsTest
[INFO] Running net.pms.util.FileUtilTest
[INFO] Tests run: 33, Failures: 0, Errors: 0, Skipped: 0, Time elapsed: 0.165 s - in net.pms.util.FileUtilTest
[INFO] Running net.pms.util.IteratorsTest
[INFO] Tests run: 1, Failures: 0, Errors: 0, Skipped: 0, Time elapsed: 0.001 s - in net.pms.util.IteratorsTest
[INFO] Running net.pms.util.jna.JNATypesTest
[INFO] Tests run: 5, Failures: 0, Errors: 0, Skipped: 0, Time elapsed: 0.017 s - in net.pms.util.jna.JNATypesTest
[INFO] Running net.pms.util.jna.macos.corefoundation.CoreFoundationTest
[WARNING] Tests run: 1, Failures: 0, Errors: 0, Skipped: 1, Time elapsed: 0.001 s - in net.pms.util.jna.macos.corefoundation.CoreFoundationTest
[INFO] Running net.pms.util.jna.macos.iokit.IOKitTest
[WARNING] Tests run: 1, Failures: 0, Errors: 0, Skipped: 1, Time elapsed: 0 s - in net.pms.util.jna.macos.iokit.IOKitTest
[INFO] Running net.pms.util.jna.macos.kernreturn.KernReturnTest
[INFO] Tests run: 4, Failures: 0, Errors: 0, Skipped: 0, Time elapsed: 0.006 s - in net.pms.util.jna.macos.kernreturn.KernReturnTest
[INFO] Running net.pms.util.LanguagesTest
[INFO] Tests run: 4, Failures: 0, Errors: 0, Skipped: 0, Time elapsed: 0 s - in net.pms.util.LanguagesTest
[INFO] Running net.pms.util.PlayerUtilTest
[INFO] Tests run: 9, Failures: 0, Errors: 0, Skipped: 0, Time elapsed: 0.006 s - in net.pms.util.PlayerUtilTest
[INFO] Running net.pms.util.RationalTest
[INFO] Tests run: 30, Failures: 0, Errors: 0, Skipped: 0, Time elapsed: 0.041 s - in net.pms.util.RationalTest
[INFO] Running net.pms.util.StringUtilTest
[INFO] Tests run: 5, Failures: 0, Errors: 0, Skipped: 0, Time elapsed: 0.001 s - in net.pms.util.StringUtilTest
[INFO] Running net.pms.util.UMSUtilsTest
[INFO] Tests run: 1, Failures: 0, Errors: 0, Skipped: 0, Time elapsed: 0.062 s - in net.pms.util.UMSUtilsTest
[INFO] Running net.pms.util.VersionTest
[INFO] Tests run: 10, Failures: 0, Errors: 0, Skipped: 0, Time elapsed: 0.004 s - in net.pms.util.VersionTest
[INFO] 
[INFO] Results:
[INFO] 
[ERROR] Errors: 
[ERROR]   DLNAMediaInfoTest.SetUPClass:76 NullPointer
[INFO] 
[ERROR] Tests run: 368, Failures: 0, Errors: 1, Skipped: 2
[INFO] 
[INFO] ------------------------------------------------------------------------
[INFO] BUILD FAILURE
[INFO] ------------------------------------------------------------------------
[INFO] Total time: 21.241 s
[INFO] Finished at: 2018-11-17T18:22:34+01:00
[INFO] ------------------------------------------------------------------------
[ERROR] Failed to execute goal org.apache.maven.plugins:maven-surefire-plugin:2.20:test (default-test) on project ums: There are test failures.
[ERROR] 
[ERROR] Please refer to E:\Workspace\UniversalMediaServer\target\surefire-reports for the individual test results.
[ERROR] Please refer to dump files (if any exist) [date]-jvmRun[N].dump, [date].dumpstream and [date]-jvmRun[N].dumpstream.
[ERROR] -> [Help 1]
[ERROR] 
[ERROR] To see the full stack trace of the errors, re-run Maven with the -e switch.
[ERROR] Re-run Maven using the -X switch to enable full debug logging.
[ERROR] 
[ERROR] For more information about the errors and possible solutions, please read the following articles:
[ERROR] [Help 1] http://cwiki.apache.org/confluence/display/MAVEN/MojoFailureException
```

**Comment by Nadahar** [comment]
The way UMS is initialized is flawed in many ways, with a lovely mix of static and instance initialization. Surefire will fork the JVM and run tests in parallel in multiple threads. All this makes the process highly unpredictable, so I suspect that you will stumble upon a lot of this type of issues.

**Comment by stasinos** [comment]
I could work around the issue with initializing PMS multiple time, by adding the ability to force a fresh PMS instance.

I now have:
 - syncResolve() failing on MacOS only
 - inconsistent container/mime reporting for Matroska (see commit message 149cb98)

The latter issue is, I guess, exactly the kind of thing these tests are meant to capture.

I am calling it a day. Please indicate if you agree with the addition of PMS.getNew() that
guarantees getting a freshly initialized PMS instance.

I will try to work out the MacOS issue on a future session, but if somebody has more experience with MacOS please help.

The next step will be to formalize the expected values in a JSON that defines the test suite, in a manner similar to the filename metadata extraction testcase. Before going there, I will need @SubJunk to make a decision on commiting sample video files in the source tree or downloading them at build time.

**Comment by SubJunk** [comment]
@stasinos nice work so far :) I think with big files we usually download them at build time, like we do with our other binaries like FFmpeg. That way we don't make everyone get them when they clone our Git repository. I made an issue for this at https://github.com/UniversalMediaServer/UniversalMediaServer/issues/1679
I am happy to work on it, or anyone can pick it up in the meantime.

File: src/main/java/net/pms/PMS.java [hunk 2]
```diff
@@ -897,6 +898,13 @@ public static PMS get() {
 		return instance;
 	}
 
+	@Nonnull
+	public static PMS getNew() {
```
Comment (SubJunk) [inline src/main/java/net/pms/PMS.java]: Maybe it would be better to name it `getNewInstance`

**Comment by SubJunk** [comment]
> The latter issue is, I guess, exactly the kind of thing these tests are meant to capture.

Yes exactly, and it's great to see this already paying off; it shows the dangers of having such a complex program without good test coverage. This is really exciting :)

File: src/main/java/net/pms/configuration/PmsConfiguration.java [hunk 3]
```diff
@@ -3186,12 +3186,20 @@ public void setSharedFolders(Vector<Vector<?>> tableVector) {
 			synchronized (sharedFoldersLock) {
 				if (!sharedFoldersRead || !sharedFolders.isEmpty()) {
 					configuration.setProperty(KEY_FOLDERS, "");
-					sharedFolders.clear();
+					if (sharedFolders==null) {
+						sharedFolders = new ArrayList<>();
```
[the remark this answers is not in the document]
Comment (stasinos) [reply src/main/java/net/pms/configuration/PmsConfiguration.java]: Done.

File: src/main/java/net/pms/PMS.java (hunk 2 above)
Comment (stasinos) [reply src/main/java/net/pms/PMS.java]: Done

**Comment by SubJunk** [comment]
@stasinos in that case let's try it, I am curious to try it too :)

**Comment by stasinos** [comment]
I have a playground here: https://github.com/stasinos/testlfs

When you clone it without LFS installed, you get what is actually commited which is files like this:
```
$ git clone git@github.com:stasinos/testlfs.git
$ cat h264-nosound.mkv 
version https://git-lfs.github.com/spec/v1
oid sha256:3c40ed93181ec7a09aa17c0f9e0b87be713cce9c33b49e3107c4d46c1fd20ba3
size 2999531
```
If you download the file from the github website, you get the actual file and not this pointer. If you install and initialize LFS, all pushing and fetching will work as usual. However, only the files that are at the tip of branches will be kept in the local .git. So the media files will be fetched, but changing them will NOT incrementally pile up as it would normally. Older versions are there on the server, and one can request that all versions are fetched locally, but the default behaviour is to only have locally what is needed for the tips of actually checked out branches.

If you install LFS and clone again stasinos/testlfs, you will now fetch the files at the tip of master:
```
$ du -s -h .git
6.9M    .git
```
There is also a commit with higher-res versions of two of the files:
```
$ git checkout ef25aeb0a1612ebe06b42056ee5caaa0c36c1a00
$ du -s -h .git
25M     .git
```
That commit is on the master branch, but not at the tip, so the objects were not fecthed before actually needed. If I move back to the tip of master, the objects will be removed from my local .git again:
```
$ git checkout ef25aeb0a1612ebe06b42056ee5caaa0c36c1a00
Filtering content: 100% (2/2), 18.01 MiB | 2.96 MiB/s, done.
Note: checking out 'ef25aeb0a1612ebe06b42056ee5caaa0c36c1a00'.
HEAD is now at ef25aeb three files
$ du -s -h .git
25M     .git
$ git checkout master
Previous HEAD position was ef25aeb three files
Switched to branch 'master'
Your branch is up to date with 'origin/master'.
$ git lfs prune
prune: 5 local object(s), 3 retained, done                                                                                       
prune: Deleting objects: 100% (2/2), done                                                                                        
$ du -s -h .git
6.9M    .git
```
(Pruning would eventually happen automatically, but I am forcing it here to demonstrate).

There is a catch: only paying Github customers can push LFS object to forks, so for a free-hosted project LFS can only be pushed directly into the original repo. So you have to do all the commits for the test case content without PRs.

What do you think @SubJunk ?

**Comment by SubJunk** [comment]
@stasinos that's a cool writeup, thanks for the info. I'm hesitant about that paid part given that this is an open source project, plus it makes it even less competitive with just using SVN like we currently do. What do you think?

**Comment by SubJunk** [comment]
@stasinos is that a deal breaker for you? i'm happy to discuss further. Also happy to give you access to our SVN repository

**Comment by SubJunk** [comment]
Season's greetings to you too, no problem we all get busy! That testcases repo looks great too.
I agree there isn't really an advantage for the SVN repo in terms of technical ability, but the fact that it already exists and we use it is nice. Also it is hosted on the UMS server which may be an advantage too depending on personal opinion :) since it's my server I like that

**Comment by SubJunk** [comment]
If you prefer LFS then use that, I don't want to block you

**Comment by stasinos** [comment]
SVN will do just fine. Please see https://github.com/stasinos/ums_testcases which reacts to Issue #1679 and, if you are happy with it, let me know how to proceed: will you commit the files to SVN? Will you give me commit access so I can do it? Will you start a git repo for these files under github.com/UniversalMediaServer? Something else? Your call.

Either way, I am not sure what should happen when using the main repo's pom to run tests. Should the user have downloaded the test content beforehand? If yes, I assume that the tests do not fail when the files are not found, correct? A failure is that the test files are there, and the tests fail.

If not, should maven download it? This can be very slow, so there should be some way to only do this when explicitly requested.

I wish you a very happy new year!

**Comment by SubJunk** [comment]
@stasinos I'm happy to give you SVN access if you want it, otherwise I can upload them. And yes I agree with you. And sorry for the extremely late reply, this didn't show up in my notifications and I only saw it now since I was clicking through the PRs :/ I hope you are still around

**Comment by SubJunk** [comment]
Continued in https://github.com/UniversalMediaServer/UniversalMediaServer/pull/1889

**Pull request closed without merging**